File: leetcode/src/main/golang/roman_to_integer.go
func romanToInt(s string) int {
    if len(s) == 0 {
        return 0
    }
    sum, cur, last := 0, 0, 0
    for _, v := range s {
        cur = convert(v)
        if last < cur {
            sum += cur - 2 * last
        } else {
            sum += cur
        }
        last = cur
    }
    return sum
}

func convert(b rune) int {
    switch b {
      case 'I':
        return 1
      case 'V':
        return 5
      case 'X':
        return 10
      case 'L':
        return 50
      case 'C':
        return 100
      case 'D':
        return 500
      case 'M':
        return 1000
      default:
        return 0
    }
}

File: leetcode/src/main/ruby/roman_to_integer.rb
# @param {String} s
# @return {Integer}
def roman_to_int(s)
    return 0 if s.length == 0
    sum, cur, last = 0, 0, 0
    s.each_char { |c|
        cur = convert(c)
        sum += last >= cur ? cur : cur - 2 * last
        last = cur
    }
    sum
end

def convert(c)
   case c
   when 'I'
       1
   when 'V'
       5
   when 'X'
       10
   when 'L'
       50
   when 'C'
       100
   when 'D'
       500
   when 'M'
       1000
   else
       0
   end
end